Support RFC links in Quick Import functionality (#11949)

AU0430 opened the pull request:
### Purpose
See #11949
This PR addresses [Issue #11949](https://github.com/JabRef/jabref/issues/11949), which requested support for importing RFC using url links in the quick import feature of JabRef.

### Changes Made
1. **Enhanced CompositeIdFetcher to Recognize Rfc URL Links**:
   - Updated the `CompositeIdFetcher` class to support RFC document URLs.
   - Added RfcId.java to parse RFC URLs (e.g., `https://www.rfc-editor.org/rfc/rfc7276.html`) and extract the RFC ID correctly, even if the URL contains a section fragment (e.g., `#section-2.2.2`).

2. **Updated Identifier Handling**:
   - Added RFC-specific parsing logic directly in the `CompositeIdFetcher` to detect URLs containing "rfc" and extract relevant RFC identifiers.
   - This change allows the `performSearchById` method to correctly identify and handle RFC URLs without needing a separate importer.

### How the Issue is Solved
- **Problem**: Users were unable to import RFC documents directly using their full URLs, especially when these URLs included section references.
- **Solution**: Modified the `CompositeIdFetcher` class to include logic that can extract RFC identifiers from URLs and proceed with the import. This allows users to directly paste RFC URLs into the quick import field, and JabRef will successfully parse and import the relevant RFC entry.
  
### Notes
- This PR only supports URLs from `https://www.rfc-editor.org`. If users encounter other RFC sources, further enhancement might be needed.
- Users can now easily use the quick import feature with full RFC URLs, which will enhance the usability of JabRef for RFC references.

Please review these changes, and let me know if there are any adjustments or additional enhancements required. Feedback is highly appreciated to ensure the solution meets user expectations effectively.

koppor commented on src/main/java/org/jabref/model/entry/identifier/RfcId.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    @Override
```
I think, neither equals nor hashCode need to be overriden.

Just remove that code.

There is also a logic error, because Id is stored lower case and the equals also converts loeer case (unnecessarily)

AU0430 commented on src/main/java/org/jabref/model/entry/identifier/RfcId.java:
```diff
@@ -0,0 +1,79 @@
+        return Optional.empty();
+    }
+
+    public boolean isValid() {
```
Hi, maybe I understand something wrong, so I don't need to match plain in this part, but only need to match URLs in this format("https://www.rfc-editor.org/rfc/" + rfc7276.html#section-2.2.2)?

AU0430 replied:
Can the modified part meet your requirements?

koppor replied:
> Hi, maybe I understand something wrong, so I don't need to match plain in this part, but only need to match URLs in this format("https://www.rfc-editor.org/rfc/" + rfc7276.html#section-2.2.2)?

You do nlt match the URL at all there. therefore I wondered.

AU0430 replied:
Because I was worried that there might be other formats of URLs, I considered just matching the format of rfc****. 
My mistake. I am really sorry for causing you trouble.

koppor replied:
The method is nowhere used. REmove.

If you would implement it correctly, you need to have to check both the URL and just the ID.

AU0430 commented:
Hello, I have modified my code style, I think there should be no problem, can you test my PR again?

AU0430 commented:
According to CI test feedback:
1. I changed the code style according to the checkstyle.xml.
1. Removed the entry RFC("rfc", "RFC", FieldProperty.IDENTIFIER) from the codebase, thereby reducing the total number of entries from 126 to 125.
2. Modified the RfcId class to extend EprintIdentifier, enhancing its functionality to integrate better with existing identifier logic.

koppor commented on src/main/java/org/jabref/model/entry/identifier/RfcId.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents an RFC identifier, which can be used to fetch bibliographic information about the RFC document.
+ * This class supports both plain RFC IDs (e.g., "rfc7276") and full URLs (e.g., "https://www.rfc-editor.org/rfc/" + rfc****.html).
+ */
+public class RfcId extends EprintIdentifier {
```
Name this class `RFC` - to be consistent with the other classes in that package.

AU0430 replied:
1. Thank you for the feedback. I have renamed the class to RFC for consistency with the other classes in the package. The class now follows the naming conventions as requested.
2. I have removed the unused method(isValid) as suggested. The relevant parsing logic has been refactored to ensure proper handling of both RFC IDs and URLs, as requested.
3. I have implemented additional test cases in RFCTest.java to verify that the new RFC class handles both plain RFC IDs (e.g., rfc7276) and URLs (e.g., https://www.rfc-editor.org/rfc/rfc7276.html).

koppor reviewed (changes requested):
Tests for RfcId class missing.

Both `rfc7276` and URLs ´https://www.rfc-editor.org/rfc/rfc7276.html` need to be supported

koppor commented on src/test/java/org/jabref/model/entry/identifier/RFCTest.java:
```diff
@@ -0,0 +1,39 @@
+class RFCTest {
+
+    @Test
+    void testParsePlainRfcId() {
```
Convert to `@ParameterizedTest`.You can use `@CsvSource`. Left: `excpected`, right: input.

Use `assertEquals(expected, new RFC(input).getNormlized()`. 

Also the next test case.

AU0430 replied:
Converted `testParsePlainRfcId` to a `@ParameterizedTest` using `@CsvSource`(done)
- Updated `testGetExternalUri` to verify the URI using `Optional.of()` for consistent checks.

koppor commented on src/test/java/org/jabref/model/entry/identifier/RFCTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    void testInvalidRfc() {
+        Optional<RFC> rfc = RFC.parse("invalidRfc");
+        assertFalse(rfc.isPresent());
```
```suggestion
        assertEquals(Optional.empty, rfc);
```

If something is wrong, the diff can be investigated.

AU0430 replied:
Adjusted `testInvalidRfc` to use `assertEquals(Optional.empty(), rfc)`.(done)

koppor commented on src/test/java/org/jabref/model/entry/identifier/RFCTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    void testGetExternalUri() {
+        RFC rfc = new RFC("rfc7276");
+        Optional<URI> uri = rfc.getExternalURI();
+        assertTrue(uri.isPresent());
+        assertEquals("https://www.rfc-editor.org/rfc/rfc7276", uri.get().toString());
```
```suggestion
        assertEquals(Optional.of("https://www.rfc-editor.org/rfc/rfc7276"), rfc.getExternalURI());
```

AU0430 replied:
Updated `testGetExternalUri` to verify the URI using `Optional.of()` for consistent checks.(done)

koppor reviewed (changes requested):
Some small comments on the test.

koppor commented on src/test/java/org/jabref/model/entry/identifier/RFCTest.java:
```diff
@@ -0,0 +1,37 @@
+
+    @ParameterizedTest
+    @CsvSource({
+            "rfc7276, rfc7276",
+            "https://www.rfc-editor.org/rfc/rfc7276.html, rfc7276"
+    })
+    void testParseValidRfcIdAndUrl(String input, String expected) {
```
Please be consistent.

assertEquals first gets expected, thus please, first `expected` then `input`

I fix it for myself. Too much context switches.

koppor reviewed (dismissed):
OK for me now.

koppor commented:
@ChengPuPu Thank you for the follow up. I was about to open the PR in my IDE and fix it for myself.

koppor commented:
BTW: No need to disable automerge. Automerge happens if the required workflows run successfully.

AU0430 commented:
@koppor Thank you very much for your patience and guidance. I have learned a lot about code management through this experience. Thank you very much.
